src/capabilities.js

```diff
@@ -0,0 +1,207 @@
+  #coreManager
+  #projectCreatorAuthCoreId
+  #ownDeviceId
+
+  /**
+   *
+   * @param {object} opts
+   * @param {import('./datatype/index.js').DataType<
+   *   import('./datastore/index.js').DataStore<'auth'>,
+   *   typeof import('./schema/project.js').roleTable,
+   *   'role',
+   *   import('@mapeo/schema').Role,
+   *   import('@mapeo/schema').RoleValue
+   * >} opts.dataType
+   * @param {import('./core-ownership.js').CoreOwnership} opts.coreOwnership
+   * @param {import('./core-manager/index.js').CoreManager} opts.coreManager
+   * @param {Buffer} opts.projectKey
+   * @param {Buffer} opts.deviceKey public key of this device
+   */
+  constructor({ dataType, coreOwnership, coreManager, projectKey, deviceKey }) {
+    this.#dataType = dataType
+    this.#coreOwnership = coreOwnership
+    this.#coreManager = coreManager
+    this.#projectCreatorAuthCoreId = projectKey.toString('hex')
+    this.#ownDeviceId = deviceKey.toString('hex')
+  }
+
+  /**
+   * Get the capabilities for device `deviceId`.
+   *
+   * @param {string} deviceId
+   * @returns {Promise<Capability>}
+   */
+  async getCapabilities(deviceId) {
+    let roleId
+    try {
+      const roleAssignment = await this.#dataType.getByDocId(deviceId)
+      roleId = roleAssignment.roleId
+    } catch (e) {
+      // The project creator will have all capabilities
+      const authCoreId = await this.#coreOwnership.getCoreId(deviceId, 'auth')
+      if (authCoreId === this.#projectCreatorAuthCoreId) {
+        return CREATOR_CAPABILITIES
+      } else {
+        return DEFAULT_CAPABILITIES[BLOCKED_ROLE_ID]
+      }
+    }
+    if (!isKnownRoleId(roleId)) {
+      return DEFAULT_CAPABILITIES[BLOCKED_ROLE_ID]
+    }
+    const capabilities = DEFAULT_CAPABILITIES[roleId]
+    return capabilities
+  }
+
+  /**
+   * Assign a role to the specified `deviceId`. Devices without an assigned role
+   * are unable to sync, except the project creator that defaults to having all
+   * capabilities. Only the project creator can assign their own role. Will
+   * throw if the device trying to assign the role lacks the `roleAssignment`
+   * capability for the given roleId
+   *
+   * @param {string} deviceId
+   * @param {keyof DEFAULT_CAPABILITIES} roleId
+   */
+  async assignRole(deviceId, roleId) {
+    let fromIndex = 0
+    let authCoreId
+    try {
+      authCoreId = await this.#coreOwnership.getCoreId(deviceId, 'auth')
+      const authCoreKey = Buffer.from(authCoreId, 'hex')
+      const authCore = this.#coreManager.getCoreByKey(authCoreKey)
+      if (authCore) {
+        await authCore.ready()
+        fromIndex = authCore.length
+      }
+    } catch {
+      // This will usually happen when assigning a role to a newly invited
+      // device that has not yet synced (so we do not yet have a replica of
+      // their authCore). In this case we want fromIndex to be 0
+    }
+    const isAssigningProjectCreatorRole =
+      authCoreId === this.#projectCreatorAuthCoreId
```

Comment:
somewhat of a nit but initially read this as assigning a role called project creator

```suggestion
    const isAssigningToProjectCreator =
      authCoreId === this.#projectCreatorAuthCoreId
```